src/core/init/stream_events_emitter/get_scheduled_events.ts

```diff
@@ -0,0 +1,77 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import config from "../../../config";
+import Manifest from "../../../manifest";
+import areSameStreamEvents from "./are_same_stream_events";
+import { IStreamEventPrivateData } from "./types";
+
+const { STREAM_EVENT_EMITTER_POLL_INTERVAL } = config;
+const boundsShift = (STREAM_EVENT_EMITTER_POLL_INTERVAL * 0.6) / 1000;
+
+/**
+ * Refresh local scheduled events list
+ */
+function getScheduledEvents(currentScheduledEvents: IStreamEventPrivateData[],
+                            manifest: Manifest): IStreamEventPrivateData[] {
+  const scheduledEvents: IStreamEventPrivateData[] = [];
+  const { periods } = manifest;
+  for (let i = 0; i < periods.length; i++) {
+    const period = periods[i];
+    const { streamEvents } = period;
+    if (streamEvents !== undefined) {
+      streamEvents.forEach((event) => {
+        const eventStart = (event.presentationTime ?? 0) / event.timescale;
+        const start = period.start + eventStart;
+        const end = event.duration !== undefined ?
+          start + (event.duration / event.timescale) :
+          undefined;
+
+        if (currentScheduledEvents.length > 0) {
+          for (let j = 0; j < currentScheduledEvents.length; j++) {
+            const currentScheduleEvent = currentScheduledEvents[j];
+            if (areSameStreamEvents(currentScheduleEvent, { id: event.id, start, end })) {
+              scheduledEvents.push(currentScheduleEvent);
+              return;
+            }
+          }
+        }
+
+        // We shift the start and end in case an event shall last less than
+        // STREAM_EVENT_EMITTER_POLL_INTERVAL
+        // Thus, we may trigger each in and out event with a tiny time shift.
+        const shouldShift = end !== undefined &&
+                            (end - start) < STREAM_EVENT_EMITTER_POLL_INTERVAL / 1000;
+        const _shiftedStart = shouldShift ? start - boundsShift :
+                                            start;
+        const _shiftedEnd = end === undefined ? undefined :
+                                               shouldShift ? (end + boundsShift) :
+                                                             end;
+        const newScheduledEvent = { start,
+                                    end,
+                                    _shiftedStart,
+                                    _shiftedEnd,
+                                    id: event.id,
+                                    _isBeingPlayed: false,
```

Comment:
What was the problem with creating a WeakMap?
Here we have a frankenstein-y object which can change state.

The `WeakMap` would have the advantage of being more explicit. Or we could even create a `StreamEvent` class instance here, where it also wouldn't be surprising if it contained mutable state.